tictactoe/adapter-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/tictactoe/adapter/scriptgen/TicTacToeScriptGenerationQueryFactory.kt

```diff
@@ -0,0 +1,115 @@
+        }
+}
+
+abstract class TicTacToeScriptGenerationQueryFactory(val queryPosition: QueryPosition) :
+    AbstractTicTacToeScriptGenerationQueryFactory() {
+
+    override fun getPlayOrder(gameName: String) = query<List<String>>(QueryReference.PlayOrder, gameName)
+    override fun getPositions(gameName: String) = query<List<Position>>(QueryReference.Positions, gameName)
+    override fun getToken(gameName: String, position: Position) =
+        query<String?>(QueryReference.Token, gameName, position)
+
+    override fun canPlayerPlaceToken(gameName: String, playerName: String, position: Position) =
+        query<Boolean>(QueryReference.CanPlayerPlaceToken, gameName, playerName, position)
+
+    private fun <T> query(reference: QueryReference, vararg args: Any?) = reference.inPosition(queryPosition)?.let {
+        TicTacToeScriptGenerationQuery<T> {
+            TicTacToeScriptingHelper.scriptifyFunction(it, *args)
+        }
+    } ?: UnsupportedQuery()
+}
+
+object TicTacToeScriptGenerationQueryQueryFactory : TicTacToeScriptGenerationQueryFactory(QueryPosition.Query)
+object TicTacToeScriptGenerationVerificationQueryFactory : TicTacToeScriptGenerationQueryFactory(QueryPosition.Verify)
+
+object TicTacToeScriptGenerationDerivedQueryFactory : AbstractTicTacToeScriptGenerationQueryFactory() {
+
+    override fun getPlayOrder(gameName: String) =
+        query<List<String>, String>(QueryReference.PlayOrder, "(insert player here)", gameName)
+
+    override fun getPositions(gameName: String) =
+        query<List<Position>, Position>(QueryReference.Positions, Position(-1, -1), gameName)
+
+    private fun <C : Collection<T>, T> query(reference: QueryReference, dummy: T, vararg args: Any?) =
+        reference.inPosition(QueryPosition.Derived)?.let {
+            ScriptGenerationQueryWithDummy(TicTacToeScriptGenerationQuery<C> {
+                TicTacToeScriptingHelper.scriptifyFunction(it, *args)
+            }, dummy)
+        } ?: UnsupportedQuery()
+}
+
+enum class QueryPosition {
+    Query, Verify, Derived
+}
+
+enum class QueryReference(
+    val inQueryPosition: KFunction<*>, val inVerificationPosition: KFunction<*>, val inDerivedPosition: KFunction<*>?
+) {
+    PlayOrder(TicTacToeQueries::getPlayOrder,
+        TicTacToeVerify::gameHasPlayOrder,
+        DerivedQueryBlock::getPlayOrder),
+    Positions(TicTacToeQueries::getPositions,
+        TicTacToeVerify::gameHasPositions,
+        DerivedQueryBlock::getPositions),
+    Token(TicTacToeQueries::getToken,
+        TicTacToeVerify::gameHasToken,
+        null),
+    CanPlayerPlaceToken(TicTacToeQueries::canPlayerPlaceToken, TicTacToeVerify::playerCanPlaceToken, null);
+
+    fun inPosition(position: QueryPosition) = when (position) {
+        QueryPosition.Query -> inQueryPosition
+        QueryPosition.Verify -> inVerificationPosition
+        QueryPosition.Derived -> inDerivedPosition
+    }
+}
```

Comment:
This might be a bit over-engineered.  I was trying to make it so that I could reuse a lot of the infrastructure between all three types of query (query-query, verify-query, and derived driver), and some of the remaining infrastructure at least between the first two.